### PR TITLE
Serialize feature

### PR DESCRIPTION
**Pull request by jxy-s**
The following PR integrates with [serde](https://serde.rs/). This enables `EventRecord` to be serialized to a variety of formats supported by `serde`.

The serialization support is implemented as a feature. Users wanting the serialization features must enable the `serde` feature for this crate. The bulk of the feature is implemented in `ser.rs`. The serialization feature is used as follows:

```Rust
use ferrisetw::schema_locator::SchemaLocator;
use ferrisetw::{EventRecord, EventSerializer};
extern crate serde_json;

fn event_callback(record: &EventRecord, schema_locator: &SchemaLocator) {
    match schema_locator.event_schema(record) {
        Err(err) => println!("Error {:?}", err),
        Ok(schema) => {
            // Generate a serializer for the record using the schema
            let ser = EventSerializer::new(record, &schema, Default::default());
            // Pass the serializer to any serde compatible serializer
            match serde_json::to_value(ser) {
                Err(err) => println!("Error {:?}", err),
                Ok(json) => println!("{}", json),
            }
        }
    }
}
```

There are two commits in the PR, the first one implements the serialization feature (https://github.com/n4r1b/ferrisetw/commit/2fae22486b960dc6eb5a0254b28634c831f24da2), the second commit (https://github.com/n4r1b/ferrisetw/commit/826855fc233dd4f03140a817b56f5eb623467b49) corrects the code based on clippy analysis `cargo clippy --tests --all-targets --all-features -- -D warnings` . I broke up the commits since the later is more or less superfluous cleanup. It might be best to review them separately.

### CODE REVIEW

**Comment by daladim** [comment]
Thanks for this awesome work!
I won't be able to review this PR this week, but I definitely will as soon as I can.

Before I read it in detail, I am curious of your use case that requires to serialize events. Why do you need it?
Also, have you investigated whether tdh (or similar Windows API) provided something to serialize a single `EVENT_RECORD`? It does make it possible to serialize them into .ETL files, but I don't know whether they have the flexibility to easily serialize single `EVENT_RECORD`s (which, I guess, is necessary to integrate with serde)

**Comment by jxy-s** [comment]
> I am curious of your use case that requires to serialize events. Why do you need it?

Integrating with various systems that consume information in different formats. I know that's broad reasoning, but an example might be - forwarding certain ETW events to Splunk as JSON. Another might be some decision engine that consumes information as flexbuffers and looks for patterns. The possibilities are pretty endless when you can transform information to standardized and self-describing formats.

> Also, have you investigated whether tdh (or similar Windows API) provided something to serialize a single EVENT_RECORD?

Microsoft doesn't support this AFAIK. And, as you mentioned it might not be possible to integrate with serde. Also, the TDH formatting APIs are notoriously slow... I appreciate that this library parses the information directly into in Rust. That's one of the reasons I chose to contribute here rather than to another project (or write my own). The implementation here is mostly complete for my needs. For now, I just needed serialization support, serde being the de facto standard for Rust - feels like a good marriage of tech :). Moving forward, payload predicate filtering would probably be become necessary for what I'm working on (already described here https://github.com/n4r1b/ferrisetw/issues/6).

**Review by n4r1b — APPROVED** [review]
This is great! Thanks @jxy-s 
LGTM! Feel free to squash/merge whenever you are ready.
Let me know if you'd like a new crate version to be published with these changes (let's raise the minor)

**Comment by jxy-s** [comment]
@n4r1b PR can be squashed and merged as it is. However, there is `1 workflow awaiting approval`. A maintainer or someone with permissions will have to approve the workflow before I am able to squash and merge.

**Comment by jxy-s** [comment]
Changes to the parsing of Unicode strings broke the DNS parsing. The code should consider when the length includes a null terminator. I'll fix it a few.

**Comment by jxy-s** [comment]
👍 ready to squash and merge, looks like I don't have permissions, however.

**Pull request merged**
